test/VaultTest.t.sol

```diff
@@ -0,0 +1,256 @@
+        uint maxDebt = vault.getMaxDebt(user);
+        vm.assume(excessDebt > maxDebt);
+        
+        vm.prank(user);
+        vm.expectRevert("Insufficient collateral!");
+        vault.borrow(excessDebt);
+    }
+
+    function testCannotWithdrawExceedingMargin(uint excessWithdrawl) public {
+        console2.log("With existing debt, user should be unable to withdraw in excess of required collateral");
+
+        uint collateralRequired = vault.getCollateralRequired(vault.debts(user));
+        uint spareDeposit = vault.deposits(user) - collateralRequired;
+        vm.assume(excessWithdrawl > spareDeposit);
+
+        vm.prank(user);
+        vm.expectRevert("Collateral unavailable!");
+        vault.withdraw(excessWithdrawl);
+    }
+
+    function testCannotBeLiquidatedAtSamePrice() public {
+        console2.log("Ceteris paribus, user should not be liquidated");
+        uint userDebt = vault.debts(user);
+        uint userDeposit = vault.deposits(user);
+
+        vault.liquidation(user);
+        assertTrue(vault.deposits(user) == userDeposit);
+        assertTrue(vault.debts(user) == userDebt);
+    }
+
+    function testRepay(uint repayAmount) public {
+        console2.log("User repays debt");
+        uint userDebt = vault.debts(user);
+
+        vm.assume(repayAmount <= userDebt);
+        vm.assume(repayAmount > 0);
+        
+
+        vm.expectEmit(true, true, false, true);
+        emit Repay(address(dai), user, repayAmount);
+
+        vm.startPrank(user);
+        dai.approve(address(vault), repayAmount);
+        vault.repay(repayAmount);   
+        vm.stopPrank();
+
+    }
+}
+
+
+abstract contract StateLiquidated is StateBorrowed {
+    function setUp() public override virtual {
+        super.setUp();
+
+        // user borrows 1/2 of maxDebt earlier
+        // modify price to cause liquidation
+        // if 1 WETH is converted to less DAI, at mkt price, Vault will be unable to recover DAI lent
+        // DAI/WETH price must appreciate for liquidation (DAI has devalued against WETH)
+        priceFeed.updateAnswer(386372840000000*10**5);
+    }
+}
+
+contract StateLiquidatedTest is StateLiquidated {
+
+    function testLiquidationOnPriceAppreciation() public {
+        console2.log("DAI/WETH price appreciated significantly; user should be liquidated");
+        uint userDebt = vault.debts(user);
+        uint userDeposit = vault.deposits(user);
+
+        vm.expectEmit(true, true, false, true);
+        emit Liquidation(address(weth),address(dai), user, userDebt, userDeposit);
+
+        vault.liquidation(user);
+        assertTrue(vault.deposits(user) == 0);
+        assertTrue(vault.debts(user) == 0);
+    }
+
+    function testOnlyOwnerCanCallLiquidate() public {
+        console2.log("Only Owner of contract can call liquidate -onlyOwner modifier");
+        vm.prank(user);
+        vm.expectRevert("Ownable: caller is not the owner");
+        vault.liquidation(user);
+    }
+}
+
```

Comment:
tests look pretty great.  i think uve covered all the states. i like the organization of the tests too.   well done

the only thing I would add is I would include some tests that the rate changes mid way,  so like -- borrow half of max, change rate, expect revert on borrowing the other half (because of rate change) borrow less and it works

Reply:
Oooh, that's a great idea. Thanks!